cli/src/commands/operation.rs

```diff
@@ -395,14 +460,541 @@ fn cmd_op_abandon(
+        }
+        from_op = merged_parents_op.unwrap();
+    }
+    let with_content_format = LogContentFormat::new(ui, command.settings())?;
+
+    let from_repo = repo_loader.load_at(&from_op)?;
+    let to_repo = repo_loader.load_at(&to_op)?;
+
+    ui.request_pager();
+    writeln!(
+        ui.stdout(),
+        "From operation {}: {}",
+        short_operation_hash(from_op.id()),
+        from_op.metadata().description,
+    )?;
+    writeln!(
+        ui.stdout(),
+        "  To operation {}: {}",
+        short_operation_hash(to_op.id()),
+        to_op.metadata().description,
+    )?;
+    writeln!(ui.stdout())?;
+
+    show_op_diff(
+        ui,
+        command,
+        &from_repo,
+        &to_repo,
+        !args.no_graph,
+        &with_content_format,
+        &args.diff_format,
+        args.patch,
+    )
+}
+
+// Merges the given `operations` into a single operation. Returns `None` if
+// there are no operations to merge.
+fn merge_operations(
+    repo_loader: &RepoLoader,
+    settings: &UserSettings,
+    mut operations: impl ExactSizeIterator<Item = OpStoreResult<Operation>>,
+) -> Result<Option<Operation>, CommandError> {
+    let num_operations = operations.len();
+    if num_operations == 0 {
+        return Ok(None);
+    }
+
+    let base_op = operations.next().transpose()?.unwrap();
+    let final_op = if num_operations > 1 {
+        let base_repo = repo_loader.load_at(&base_op)?;
+        let mut tx = base_repo.start_transaction(settings);
+        for other_op in operations {
+            let other_op = other_op?;
+            tx.merge_operation(other_op)?;
+            tx.mut_repo().rebase_descendants(settings)?;
+        }
+        let tx_description = format!("merge {} operations", num_operations);
+        let merged_repo = tx.write(tx_description).leave_unpublished();
+        merged_repo.operation().clone()
+    } else {
+        base_op
+    };
+
+    Ok(Some(final_op))
+}
+
+// Computes and shows the differences between two operations, using the given
+// `Repo`s for the operations.
+#[allow(clippy::too_many_arguments)]
+fn show_op_diff(
+    ui: &mut Ui,
+    command: &CommandHelper,
+    from_repo: &Arc<ReadonlyRepo>,
+    to_repo: &Arc<ReadonlyRepo>,
+    show_graph: bool,
+    with_content_format: &LogContentFormat,
+    diff_format_args: &DiffFormatArgs,
+    patch: bool,
+) -> Result<(), CommandError> {
+    let diff_workspace_command =
+        command.for_loaded_repo(ui, command.load_workspace()?, to_repo.clone())?;
+    let diff_renderer = diff_workspace_command.diff_renderer_for_log(diff_format_args, patch)?;
+
+    // Create a new transaction starting from `to_repo`.
+    let mut workspace_command =
+        command.for_loaded_repo(ui, command.load_workspace()?, to_repo.clone())?;
+    let mut tx = workspace_command.start_transaction();
+    // Merge index from `from_repo` to `to_repo`, so commits in `from_repo` are
+    // accessible.
+    tx.mut_repo().merge_index(from_repo);
```

Comment:
Perhaps, we can use the current (head) `repo.index()`? It should contain all historical commits.

Reply:
OTOH, doing it the current way means it works even if `--from` or `--to` point to an operation that was never "published" (made an op head).

Reply:
> doing it the current way means it works even if `--from` or `--to` point to an operation that was never "published" (made an op head).

I don't know if such functionality is needed, but there might be `(current_repo, from_repo, to_repo)` tuple instead of `(from_repo, to_repo)`. The `current_repo` provides a context (or visible heads + index) where repo-level templates will be evaluated.

I'm not sure if `current_repo` should be selected by `--at-op` or `--to`, but `jj op show` doesn't need to merge indexes. `jj op diff` will have to merge if visible heads are controlled by `--to`, not by `--at-op`.

Reply:
I've updated the function to use `from_repo`, `to_repo`, and a separate repo for merging (if required). There's also no longer any merging of indices for `jj op show`.